src/main/java/com/aws/greengrass/device/session/SessionManager.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.device.session;
+
+import com.aws.greengrass.device.exception.AuthenticationException;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.util.Pair;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Singleton class for managing AuthN and AuthZ session.
+ */
+public class SessionManager {
+    private static final Logger logger = LogManager.getLogger(SessionManager.class);
+    public static final String SESSION_ID = "SessionId";
+
+    // There are two types of Session IDs defined here. Internal session IDs
+    // are derived by hashing device credentials. This is done so that we can
+    // avoid creating multiple sessions for the same device. Only a single session
+    // for a given set of device credentials is allowed at a given time.
+    // However, Session IDs should not be predictable. Here, we create a separate
+    // external Session ID which can be returned to the IPC client. The external
+    // Session ID is stored along with the Session itself so that we can clean up
+    // duplicate external Session IDs
+    @Getter(AccessLevel.PACKAGE)
+    private final Map<String, Pair<String,Session>> sessionMap = new ConcurrentHashMap<>();
+    private final Map<String, String> externalToInternalSessionMap = new ConcurrentHashMap<>();
+
+    /**
+     * find session by id.
+     *
+     * @param sessionId session identifier
+     * @return session or null
+     */
+    public Session findSession(String sessionId) {
+        String internalSessionId = externalToInternalSessionMap.get(sessionId);
+        if (internalSessionId == null) {
+            return null;
+        }
+        return sessionMap.getOrDefault(internalSessionId, null).getRight();
```

Comment:
get or default is useless here since if it doesn't exist it will return null already. And you'll then cause a NPE.

Reply:
Yep, good catch